drivers/crypto/crypto_stm32.c

```diff
@@ -0,0 +1,400 @@
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t vec[BLOCK_LEN_WORDS];
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)vec, sizeof(vec), iv, BLOCK_LEN_BYTES);
+	session->config.pInitVect = vec;
+
+	/* Prefix IV to ciphertext */
+	memcpy(pkt->out_buf, iv, 16);
+
+	ret = do_encrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf + 16);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len + 16;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_cbc_decrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t vec[BLOCK_LEN_WORDS];
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)vec, sizeof(vec), iv, BLOCK_LEN_BYTES);
+	session->config.pInitVect = vec;
+
+	ret = do_decrypt(ctx, pkt->in_buf + 16, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len - 16;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_ctr_encrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t ctr[BLOCK_LEN_WORDS] = {0};
+	int ivlen = ctx->keylen - (ctx->mode_params.ctr_info.ctr_len >> 3);
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)ctr, sizeof(ctr), iv, ivlen);
+	session->config.pInitVect = ctr;
+
+	ret = do_encrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_ctr_decrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t ctr[BLOCK_LEN_WORDS] = {0};
+	int ivlen = ctx->keylen - (ctx->mode_params.ctr_info.ctr_len >> 3);
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)ctr, sizeof(ctr), iv, ivlen);
+	session->config.pInitVect = ctr;
+
+	ret = do_decrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_get_unused_session_index(void)
+{
+	int i;
+
+	for (i = 0; i < CRYPTO_MAX_SESSION; i++) {
+		if (!crypto_stm32_sessions[i].in_use) {
+			crypto_stm32_sessions[i].in_use = true;
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static int crypto_stm32_session_setup(struct device *dev,
+				      struct cipher_ctx *ctx,
+				      enum cipher_algo algo,
+				      enum cipher_mode mode,
+				      enum cipher_op op_type)
+{
+	int ctx_idx;
+	struct crypto_stm32_session *session;
+
+	if (ctx->flags & ~(CRYP_SUPPORT)) {
+		LOG_ERR("Unsupported flag");
+		return -EINVAL;
+	}
+
+	if (algo != CRYPTO_CIPHER_ALGO_AES) {
+		LOG_ERR("Unsupported algo");
+		return -EINVAL;
+	}
+
+	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
+	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
+	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
+		LOG_ERR("Unsupported mode");
+		return -EINVAL;
+	}
+
+	if ((ctx->keylen != 16U) &&
+	    (ctx->keylen != 24U) &&
```

Comment:
Note that a 192 bits key is optional and may not be supported by all HW crypto IP.
Unfortunately, there is currently no HAL API to know this capability on the runtime. It should be statically managed.
This implementation should be fine for F4 but this should be updated to have a code more generic addressing all the ST families.

Reply:
Thanks for pointing this out. I'll add a comment to this line, so this will later be configurable via a device tree option.